### PR TITLE
Project Windex: docs edition

### PR DESCRIPTION
**Pull request by boblannon**
### Context

This PR is focused on updating and improving source code documentation. The goal is total coverage of the source as it currently exists.

JSDoc comment blocks throughout the codebase to make carmen's source code more approachable. Carmen is very complex, and the goal is to make it easier to understand the context for a given function, with some indication of typical usage.

In addition, the [documentation.js](http://documentation.js.org/) library will be used to generate a markdown version of the parts of the carmen API that are public. We're taking "public" to mean the parts of the API that are typically engaged with when using carmen as a geocoding library (rather than developing on carmen itself).

### Next Steps

@boblannon
- [x] documentation.js setup
- [x] add "build-docs" command
- [x] Indexer js docs
- [ ] Text-processing js docs

@aarthykc 
- [x] geocoder js docs
- [x] Bin js docs
- [x] Template for tests in the test/README.md

cc @mapbox/geocoding-gang

### CODE REVIEW

**Comment by aarthykc** [comment]
@boblannon i branched off this branch for working through the test refactor + geocoder functions. Going to keep pushing to `windex-geocode`

**Comment by boblannon** [comment]
Some notes on the work so far:

### Factoring out anonymous functions

The code in `Geocoder` involved a 135-line anonymous function that handles a lot of the important logic around instantiating and configuring the actual indexes that are used by carmen.  In order to make this a bit more document-able, we factored this out into a separate function.  Where possible, we'd like to continue replacing anonymous functions with named ones, especially when they do such important work.

### Tilelive support

A central component of carmen is the interface to the actual indexed data. In earlier docs, this was referred to as something that typically inherits from a tilelive Tilesource/Tilesink. The [API requirement doc in the tilelive repo](https://github.com/mapbox/tilelive/blob/master/API.md) does a good job of describing most of what the interface needs to look like.

Describing the tilelive spec doesn't give enough of an explanation for the interface in question, though. In fact, carmen only supports a small subset of tilelive backend modules: [tilelive-s3](https://github.com/mapbox/tilelive-s3), [node-mbtiles](https://github.com/mapbox/node-mbtiles), and `MemSource`, defined in `carmen/lib/api-mem.js`.

The reason that other tilelive modules won't work with carmen is that carmen has its own requirements for these interfaces.  In the modules that are supported, these requirements have been added on the modules themselves. See, for example, https://github.com/mapbox/node-mbtiles/commit/51c598ddaca615ed3fea08f50d343de6a6d43053.

### Custom type: `CarmenSource`

Following on from the previous section, the full specification for a source interface doesn't have a dedicated specification in carmen's code, even as an abstract interface template. Adding something like that might be a good idea, but is out of scope for this PR. For documentation purposes, however, we're using JSDoc's `@typedef` tag to document a `CarmenSource` type (see https://github.com/mapbox/carmen/commit/d694343f767bf5a4f9d026c861e4d93c3feb2ee3).  This gives us one piece of documentation to point to when making reference to this interface in the `@param`s of other functions.

**Comment by aarthykc** [comment]
@boblannon I added a [`README.md` for tests](https://github.com/mapbox/carmen/blob/d76a8387b1147f9353725467c5a54103b292e71e/test/README.md), do you mind taking a look?

**Comment by aarthykc** [comment]
👋 @samely and @ingalls, @boblannon and I are ready for this PR to be reviewed. This round of code reorganisation has been towards trying to get a better understanding of what Carmen does and improving documentation for indexer parts of the code as well as forward and reverse geocoding.

#### What we'd like you to review:
- [ ] That the indexer docs and the key parts of geocoding like phrasematch, spatialmatch, verifymatch make sense 
- [ ] The api docs make sense
- [ ] @boblannon set up `documentation.js` and added conditions for tests to [check whether jsdocs are written correctly](https://github.com/mapbox/carmen/pull/691/files#diff-df39304d828831c44a2b9f38cd45289cR20) 
- [ ] Test [README.md](https://github.com/mapbox/carmen/pull/691/files#diff-c30ff793110b0960f76bb3854af6f5a8) to make it easier for folks to write tests and tips to navigate through carmen
- [ ] Docs reorganization

File: docs/api.md [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+-   `options` **[Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)** options
+    -   `options.tokens` **[PatternReplaceMap](#patternreplacemap)** A [PatternReplaceMap](#patternreplacemap) used to perform custom string replacement at index and query time.
+    -   `options.geocoder_inverse_tokens` **[Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)&lt;[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String), ([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) \| [Function](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements/function))>** for reversing abbreviations. Replace key with a stipulated string value or pass it to a function that returns a string. see [text-processsing](./text-processing.md) for details.
```
Comment (samely) [inline docs/api.md]: `text-processing` link is not working

File: docs/api.md [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+
+### CarmenSource
+
+[index.js:56-338](https://github.com/mapbox/carmen/blob/d76a8387b1147f9353725467c5a54103b292e71e/index.js#L23-L40 "Source code on GitHub")
+
+An interface to the underlying data that a [Geocoder](#geocoder) instance is indexing and querying. In addition to the properties described below, instances must satisfy interface requirements for `Tilesource` and `Tilesink`. See tilelive [API Docs](https://github.com/mapbox/tilelive/blob/master/API.md) for more info. Currently, carmen supports the following tilelive modules:
+
+-   [tilelive-s3](https://github.com/mapbox/tilelive-s3)
+-   [node-mbtiles](https://github.com/mapbox/node-mbtiles)
+-   [MemSource](MemSource)
```
Comment (samely) [inline docs/api.md]: `MemSource` link is not working

File: docs/api.md [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+-   `geocoderDataIterator` **function (type)** custom method for iterating over documents in the source.
+-   `getIndexableDocs` **function (pointer, callback)** get documents needed to create a forward geocoding datasource. `pointer` is an optional object that has different behavior depending on the implementation. It is used to indicate the state of the database, similar to a cursor, and can allow pagination, limiting, etc. `callback` is called with `(error, documents, pointer)` in which `documents` is a list of objects.
+
+### PatternReplaceMap
+
+[lib/util/token.js:253-268](https://github.com/mapbox/carmen/blob/d76a8387b1147f9353725467c5a54103b292e71e/lib/util/token.js#L220-L245 "Source code on GitHub")
+
+A mapping from patterns (keys) to replacements (values).
+
+-   The patterns are used in [createGlobalReplacer](createGlobalReplacer) to create `XRegExp`s (case-insensitive)
```
Comment (samely) [inline docs/api.md]: `createGlobalReplacer` link is not working

File: docs/api.md [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+
+| if a shard value...              | `data1`     | `data2`     |
+| -------------------------------- | ----------- | ----------- |
+| ...exists in `from1` and `from2` | `Feature`   | `Feature`   |
+| ...only exists in `from1`        | `Feature`   | `undefined` |
+| ...only exists in `from2`        | `undefined` | `Feature`   |
+
+#### RocksDB grid and freq caches
+
+see [RocksDBCache#merge](RocksDBCache#merge).
```
Comment (samely) [inline docs/api.md]: `RocksDBCache#merge` link is not working

File: docs/README.md [hunk 5]
```diff
@@ -1,31 +1,10 @@
...
-| lib/indexer    | TODO                                                    |
-| lib/util       | algorithmically simple utilities                        |
-| lib/pure       | pure algorithms                                         |
-| scripts        | common tasks, runnable from the command line            |
-| test           | tests                                                   |
-| bench          | bench...es?                                             |
+- [How Carmen Works](./how-carmen-works.md): high level explanation of carmen's design. Explains how queries are processed, and how results are ranked.
+- [Data Sources](./data-sources.md): how to prepare and store data for indexing
+- [Index Structure](./index-structure.md): details about index storage and cacheing
+- [Text Processing](./text-processing.md): how text documents and queries are normalized
```
Comment (samely) [inline docs/README.md]: `Text Processing` link is not working

**Comment by samely** [comment]
@boblannon @aarthykc such a great job with the documentation :notebook:  :heart_eyes: :clap: Super useful to understand how `carmen` works. 

I have some thoughts about the main `README.md`:

:thought_balloon: How it starts looks good, but as long as you are reading, it turns from `carmen` to `how to make documentation`. The idea of the documentation part is how to contribute regarding documentation? 

:thought_balloon: Links for `API docs` and `example` is being repeated in two sections in the `README.md` and I think it would easy to just have it in the API segment and create a new doc that could contain the [**API documentation style guide**](https://github.com/mapbox/carmen/tree/windex-docs#api-documentation-style-guide) and be linked to the documentation segment.

:thought_balloon: In `general documentation` segment would be good to include what :file_folder:  `carmen/docs/README.md`  describes.

----

_:spiral_notepad: one link more... [`memsource link`](https://github.com/mapbox/carmen/tree/windex-docs/example#initialize-geocoder) is not working_

File: docs/data-sources.md [hunk 6]
```diff
@@ -1,20 +1,68 @@
 # Data Sources
 
+Carmen was designed with the goal of being agnostic to how its underlying data is stored. This goal has been acheived to some degree, but there's still work to be done.
```
Comment (ingalls) [inline docs/data-sources.md]: Hmm carmen is very specific about how the data is stored (GeoJSON) but not about what data is used as sources for the geocoder, is this what you mean? Ie Locations of Bears Habitats, Climbing Areas, etc.

File: docs/data-sources.md [hunk 7]
```diff
@@ -1,20 +1,68 @@
...
+| geocoder_name           | Optional + advanced. A string to use instead of the provided config index id/key allowing multiple indexes to be treated as a single "logical" index.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
```
Comment (ingalls) [inline docs/data-sources.md]: Are we missing `geocoder_stack`, I know there was talk of renaming it but don't see something here that could have replaced it

File: docs/data-sources.md [hunk 8]
```diff
@@ -1,20 +1,68 @@
...
+| `carmen:score`                | Optional. A float or integer to sort equally relevant results by. Higher values appear first. Docs with negative scores can contribute to a final result but are only returned if included in matches of a forward search query.                                                                                               |
+| `carmen:addressnumber`        | Optional. Used with `geocoder_address`. An array of addresses corresponding to the order of their geometries in the `GeometryCollection`                                                                                                                                                                                       |
+| `carmen:types`                | Optional. An array of types associating this feature with one or more feature classes defined by the source-level `geocoder_type` key. By setting multiple types a feature can move between various feature levels depending on the query and results. If omitted, defaults to the `geocoder_type` set by the feature's index. |
```
Comment (ingalls) [inline docs/data-sources.md]: Geocoder stack is also missing here.

File: docs/data-sources.md (hunk 6 above)
Comment (boblannon) [reply docs/data-sources.md]: hm, yeah, it's a little laconic.  the point was that although it seems to accept any tilelive source, in reality it only accepts a few.  also, it won't be able to accept any more until they've implemented carmen-specific methods

**Comment by boblannon** [comment]
thanks @samely and @ingalls !  fixed the issues you pointed out.

**Comment by apendleton** [comment]
For some reason I thought it would auto-notify on this PR too, but since it didn't: would you folks mind pulling in https://github.com/mapbox/carmen/pull/694 before this goes out?

**Comment by boblannon** [comment]
integration tests look good.  going to merge and bump the version.

**Pull request merged**